// @a2ui/web_core/dist/src/v0_9/state/surface-components-model.js
import { EventEmitter } from '../common/events.js';
/**
 * Manages the collection of components for a specific surface.
 */
export class SurfaceComponentsModel {
    constructor() {
        this.components = new Map();
        this._onCreated = new EventEmitter();
        this._onDeleted = new EventEmitter();
        /** Fires when a new component is added to the model. */
        this.onCreated = this._onCreated;
        /** Fires when a component is removed, providing the ID of the deleted component. */
        this.onDeleted = this._onDeleted;
    }
    get(id) {
        return this.components.get(id);
    }
    addComponent(component) {
        if (this.components.has(component.id)) {
            throw new Error(`Component with id '${component.id}' already exists.`);
        }
        this.components.set(component.id, component);
        this._onCreated.emit(component);
    }
    removeComponent(id) {
        const component = this.components.get(id);
        if (component) {
            this.components.delete(id);
            component.dispose();
            this._onDeleted.emit(id);
        }
    }
    dispose() {
        for (const component of this.components.values()) {
            component.dispose();
        }
        this.components.clear();
        this._onCreated.dispose();
        this._onDeleted.dispose();
    }
}
