// @a2ui/web_core/dist/src/v0_9/state/surface-group-model.js
import { EventEmitter } from '../common/events.js';
/**
 * The root state model for the A2UI system.
 * Manages the collection of active surfaces.
 */
export class SurfaceGroupModel {
    constructor() {
        this.surfaces = new Map();
        this.surfaceUnsubscribers = new Map();
        this._onSurfaceCreated = new EventEmitter();
        this._onSurfaceDeleted = new EventEmitter();
        this._onAction = new EventEmitter();
        /** Fires when a new surface is added. */
        this.onSurfaceCreated = this._onSurfaceCreated;
        /** Fires when a surface is removed. */
        this.onSurfaceDeleted = this._onSurfaceDeleted;
        /** Fires when an action is dispatched from ANY surface in the group. */
        this.onAction = this._onAction;
    }
    addSurface(surface) {
        if (this.surfaces.has(surface.id)) {
            console.warn(`Surface ${surface.id} already exists. Ignoring.`);
            return;
        }
        this.surfaces.set(surface.id, surface);
        // Subscribe to surface actions and propagate
        const sub = surface.onAction.subscribe((action) => this._onAction.emit(action));
        this.surfaceUnsubscribers.set(surface.id, sub);
        this._onSurfaceCreated.emit(surface);
    }
    deleteSurface(id) {
        const surface = this.surfaces.get(id);
        if (surface) {
            const sub = this.surfaceUnsubscribers.get(id);
            if (sub) {
                sub.unsubscribe();
                this.surfaceUnsubscribers.delete(id);
            }
            this.surfaces.delete(id);
            surface.dispose();
            this._onSurfaceDeleted.emit(id);
        }
    }
    getSurface(id) {
        return this.surfaces.get(id);
    }
    dispose() {
        for (const id of Array.from(this.surfaces.keys())) {
            this.deleteSurface(id);
        }
        this._onSurfaceCreated.dispose();
        this._onSurfaceDeleted.dispose();
        this._onAction.dispose();
    }
}
