// @a2ui/web_core/dist/src/v0_9/processing/message-processor.js
import { SurfaceModel } from '../state/surface-model.js';
import { SurfaceGroupModel } from '../state/surface-group-model.js';
import { ComponentModel } from '../state/component-model.js';
/**
 * The central processor for A2UI messages.
 * @template T The concrete type of the ComponentApi.
 */
export class MessageProcessor {
    /**
     * @param catalogs A list of available catalogs.
     * @param actionHandler A global handler for actions from all surfaces.
     */
    constructor(catalogs, actionHandler) {
        this.catalogs = catalogs;
        this.actionHandler = actionHandler;
        this.model = new SurfaceGroupModel();
        this.model.onAction.subscribe(this.actionHandler);
    }
    /**
     * Subscribes to surface creation events.
     */
    onSurfaceCreated(handler) {
        return this.model.onSurfaceCreated.subscribe(handler);
    }
    /**
     * Subscribes to surface deletion events.
     */
    onSurfaceDeleted(handler) {
        return this.model.onSurfaceDeleted.subscribe(handler);
    }
    processMessages(messages) {
        for (const message of messages) {
            this.processMessage(message);
        }
    }
    processMessage(message) {
        if (message.createSurface) {
            this.processCreateSurfaceMessage(message);
            return;
        }
        const updateTypes = ['updateComponents', 'updateDataModel', 'deleteSurface'].filter(k => message[k]);
        if (updateTypes.length > 1) {
            console.warn(`Message contains multiple update types: ${updateTypes.join(', ')}. Ignoring.`);
            return;
        }
        if (message.deleteSurface) {
            this.processDeleteSurfaceMessage(message);
            return;
        }
        if (message.updateComponents) {
            this.processUpdateComponentsMessage(message);
            return;
        }
        if (message.updateDataModel) {
            this.processUpdateDataModelMessage(message);
            return;
        }
    }
    processCreateSurfaceMessage(message) {
        const payload = message.createSurface;
        const { surfaceId, catalogId, theme } = payload;
        // Find catalog
        const catalog = this.catalogs.find(c => c.id === catalogId);
        if (!catalog) {
            console.warn(`Catalog not found: ${catalogId}`);
            return;
        }
        if (this.model.getSurface(surfaceId)) {
            console.warn(`Surface ${surfaceId} already exists. Ignoring.`);
            return;
        }
        const surface = new SurfaceModel(surfaceId, catalog, theme);
        this.model.addSurface(surface);
    }
    processDeleteSurfaceMessage(message) {
        const payload = message.deleteSurface;
        if (!payload.surfaceId)
            return;
        this.model.deleteSurface(payload.surfaceId);
    }
    processUpdateComponentsMessage(message) {
        const payload = message.updateComponents;
        if (!payload.surfaceId)
            return;
        const surface = this.model.getSurface(payload.surfaceId);
        if (!surface) {
            console.warn(`Surface not found for message: ${payload.surfaceId}`);
            return;
        }
        for (const comp of payload.components) {
            const { id, component, ...properties } = comp;
            const existing = surface.componentsModel.get(id);
            if (existing) {
                if (component && component !== existing.type) {
                    // Recreate component if type changes
                    surface.componentsModel.removeComponent(id);
                    const newComponent = new ComponentModel(id, component, properties);
                    surface.componentsModel.addComponent(newComponent);
                }
                else {
                    existing.properties = properties;
                }
            }
            else {
                if (!component) {
                    console.warn(`Cannot create component ${id} without a type.`);
                    continue;
                }
                const newComponent = new ComponentModel(id, component, properties);
                surface.componentsModel.addComponent(newComponent);
            }
        }
    }
    processUpdateDataModelMessage(message) {
        const payload = message.updateDataModel;
        if (!payload.surfaceId)
            return;
        const surface = this.model.getSurface(payload.surfaceId);
        if (!surface) {
            console.warn(`Surface not found for message: ${payload.surfaceId}`);
            return;
        }
        const path = payload.path || '/';
        const value = payload.value;
        surface.dataModel.set(path, value);
    }
}
